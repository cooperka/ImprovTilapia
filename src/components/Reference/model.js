import { initPersistedStore } from '../../storageUtils';

const STORAGE_KEY = 'ReferenceSettingsModel';

export class ReferenceSettingsModel {
  constructor() {
    initPersistedStore(this, STORAGE_KEY).catch((error) => {
      console.error('Failed to initPersistedStore:', error);
    });
  }
}
