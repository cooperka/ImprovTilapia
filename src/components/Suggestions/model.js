import { observable } from 'mobx';

import { initPersistedStore } from '../../storageUtils';

const STORAGE_KEY = 'SuggestionsSettingsModel';

export class SuggestionsSettingsModel {
  @observable
  shouldAutoSuggest = false;

  @observable
  autoSuggestIntervalSeconds = 10;

  constructor() {
    initPersistedStore(this, STORAGE_KEY).catch((error) => {
      console.error('Failed to initPersistedStore:', error);
    });
  }
}
