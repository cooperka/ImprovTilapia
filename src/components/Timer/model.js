import { observable } from 'mobx';

import { initPersistedStore } from '../../storageUtils';

const STORAGE_KEY = 'TimerSettingsModel';

export class TimerSettingsModel {
  @observable
  shouldIncreaseBrightness = true;

  constructor() {
    initPersistedStore(this, STORAGE_KEY).catch((error) => {
      console.error('Failed to initPersistedStore:', error);
    });
  }
}
