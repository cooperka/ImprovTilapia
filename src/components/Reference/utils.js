import { primary, secondary, other } from './lists/vocab';

export const references = [
  {
    name: 'Primary building blocks',
    iconName: 'dictionary',
    items: primary,
  },
  {
    name: 'Secondary building blocks',
    iconName: 'dictionary',
    items: secondary,
  },
  {
    name: 'Additional concepts (WIP)',
    iconName: 'dictionary',
    items: other,
  },
];
