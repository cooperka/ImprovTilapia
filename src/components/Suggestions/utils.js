import { relationships } from '../../suggestions/relationships';
import { occupations } from '../../suggestions/occupations';
import {
  clearEmotions,
  groupedEmotions,
  emotionalDiads,
} from '../../suggestions/emotions';
import { locations } from '../../suggestions/locations';

const getRandomThing = (list) => () => {
  return list[Math.floor(Math.random() * list.length)];
};

export const suggestionTypes = {
  RELATIONSHIP: 'RELATIONSHIP',
  OCCUPATION: 'OCCUPATION',
  EMOTION_SHORT: 'EMOTION_SHORT',
  EMOTION_FULL: 'EMOTION_FULL',
  EMOTIONAL_DIAD: 'EMOTIONAL_DIAD',
  LOCATION: 'LOCATION',
};

export const suggestions = {
  [suggestionTypes.RELATIONSHIP]: {
    name: 'Relationship',
    getRandomThing: getRandomThing(relationships),
  },
  [suggestionTypes.OCCUPATION]: {
    name: 'Occupation',
    getRandomThing: getRandomThing(occupations),
  },
  [suggestionTypes.EMOTION_SHORT]: {
    name: 'Emotion (main)',
    getRandomThing: getRandomThing(clearEmotions),
  },
  [suggestionTypes.EMOTION_FULL]: {
    name: 'Emotion (full)',
    getRandomThing: getRandomThing(groupedEmotions),
  },
  [suggestionTypes.EMOTIONAL_DIAD]: {
    name: 'Emotional diad',
    getRandomThing: getRandomThing(emotionalDiads),
  },
  [suggestionTypes.LOCATION]: {
    name: 'Location',
    getRandomThing: getRandomThing(locations),
  },
};
