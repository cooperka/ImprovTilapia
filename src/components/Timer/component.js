import { Duration } from 'luxon';
import React, { Component } from 'react';
import { StyleSheet, View, Animated } from 'react-native';
import { Constants, KeepAwake, Brightness, Permissions } from 'expo';
import { Button, FAB } from 'react-native-paper';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { observer, inject } from 'mobx-react/native';

import theme from '../App/paperTheme';

import FloatingNav from '../FloatingNav/component';

const getFabIcon = (name) => () => (
  <MaterialCommunityIcons
    style={styles.icon}
    name={name}
    size={36}
    color="white"
  />
);
const PlayIcon = getFabIcon('play');
const PauseIcon = getFabIcon('pause');

function formatTime(seconds) {
  return Duration.fromObject({ seconds }).toFormat('m:ss');
}

function formatTimeCompact(seconds) {
  return Duration.fromObject({ seconds }).toFormat(
    seconds % 60 === 0 ? 'm' : ':ss',
  );
}

@inject('timerSettings')
@observer
class Timer extends Component {
  static navigationOptions = ({ navigation }) => ({
    tabBarIcon: ({ tintColor }) => (
      <MaterialCommunityIcons name="timer" color={tintColor} size={20} />
    ),
    tabBarVisible: !navigation.getParam('isFullscreen'),
  });

  state = {
    seconds: 0,
    isRunning: false,
    lastSetTime: 0,
    width: null,
    originalBrightness: undefined,
    flashValue: new Animated.Value(1),
  };

  componentWillMount() {
    const {
      timerSettings: { shouldIncreaseBrightness },
    } = this.props;
    if (shouldIncreaseBrightness) {
      Permissions.askAsync(Permissions.SYSTEM_BRIGHTNESS);
    }
  }

  componentDidMount() {
    this.timer = setInterval(this.tick, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }

  tick = () => {
    const { seconds, isRunning } = this.state;

    if (seconds <= 0) {
      if (isRunning) {
        this.startFlashAnim();
      }
      this.setState({ seconds: 0, isRunning: false });
    }

    // TODO: Save actual milliseconds instead of this short-term hack.
    this.setState(({ seconds, isRunning }) => ({
      seconds: seconds - (isRunning ? 1 : 0),
    }));
  };

  setMaxBrightness = async () => {
    const {
      timerSettings: { shouldIncreaseBrightness },
    } = this.props;
    if (!shouldIncreaseBrightness) {
      return;
    }

    const { status } = await Permissions.getAsync(
      Permissions.SYSTEM_BRIGHTNESS,
    );

    if (status === 'granted') {
      const originalBrightness = await Brightness.getBrightnessAsync();
      this.setState({ originalBrightness });
      Brightness.setBrightnessAsync(1.0);
    }
  };

  undoBrightness = async () => {
    const {
      timerSettings: { shouldIncreaseBrightness },
    } = this.props;
    if (!shouldIncreaseBrightness) {
      return;
    }

    const { originalBrightness } = this.state;
    const { status } = await Permissions.getAsync(
      Permissions.SYSTEM_BRIGHTNESS,
    );

    if (status === 'granted' && originalBrightness !== undefined) {
      Brightness.setBrightnessAsync(originalBrightness);
    }
  };

  startFlashAnim = () => {
    const { flashValue } = this.state;

    Animated.loop(
      Animated.sequence([
        Animated.timing(flashValue, {
          toValue: 0.2,
          duration: 500,
          useNativeDriver: true,
          isInteraction: false,
        }),
        Animated.timing(flashValue, {
          toValue: 1,
          duration: 500,
          useNativeDriver: true,
          isInteraction: false,
        }),
      ]),
      {
        iterations: 5,
      },
    ).start(() => this.undoBrightness());
  };

  stopFlashAnim = () => {
    const { flashValue } = this.state;
    Animated.timing(flashValue).stop();
    flashValue.setValue(1);
  };

  handleToggleTimer = () => {
    const { isRunning } = this.state;
    const shouldPlay = !isRunning;

    if (shouldPlay) {
      this.setMaxBrightness();
    }

    this.setState({ isRunning: shouldPlay });
  };

  handleReset = () => {
    const { seconds, lastSetTime } = this.state;

    this.stopFlashAnim();
    this.setState({
      seconds: lastSetTime === seconds ? 0 : lastSetTime,
      isRunning: false,
      lastSetTime: 0,
    });
  };

  handleAddTime = (additionalSeconds) => () => {
    this.stopFlashAnim();
    this.setState(({ seconds }) => {
      const newTime = seconds + additionalSeconds;
      return { seconds: newTime, lastSetTime: newTime };
    });
  };

  handleLayoutChange = ({
    nativeEvent: {
      layout: { width, height },
    },
  }) => {
    const { navigation } = this.props;
    navigation.setParams({ isFullscreen: width > height });

    this.setState({ width });
  };

  render() {
    const { navigation } = this.props;
    const { isRunning, width, flashValue } = this.state;

    return (
      <View style={styles.container} onLayout={this.handleLayoutChange}>
        <KeepAwake />

        <FloatingNav navigation={navigation} />

        <View style={styles.timeContainer}>
          <Animated.Text
            id="timeRemaining"
            style={[
              styles.time,
              { fontSize: width / 4.0, opacity: flashValue },
            ]}
          >
            {formatTime(this.state.seconds)}
          </Animated.Text>
        </View>

        <View style={styles.buttonsContainer}>
          <View style={styles.buttonsRow}>
            <Button style={styles.button} onPress={this.handleReset}>
              Reset
            </Button>
            <FAB
              key={`playPause_${isRunning}`}
              style={styles.fab}
              icon={isRunning ? PauseIcon : PlayIcon}
              onPress={this.handleToggleTimer}
            />
            <Button style={styles.button} onPress={this.handleAddTime(10)}>
              +0:10
            </Button>
          </View>

          <View style={styles.buttonsRow}>
            {[1, 5, 10, 30]
              .map((num) => num * 60)
              .map((seconds) => (
                <Button
                  key={seconds}
                  id={seconds}
                  style={styles.button}
                  onPress={this.handleAddTime(seconds)}
                >
                  {`+${formatTimeCompact(seconds)}`}
                </Button>
              ))}
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
  timeContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: Constants.statusBarHeight,
    marginBottom: 8,
  },
  time: {
    color: '#f00',
  },
  buttonsContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    maxWidth: 500,
    maxHeight: 180,
  },
  buttonsRow: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginTop: 24,
    marginHorizontal: 16,
  },
  button: {
    flex: 1,
    marginHorizontal: 16,
  },
  fab: {
    backgroundColor: theme.colors.primary,
    marginHorizontal: 16,
  },
  icon: {
    textAlign: 'center',
    marginTop: 'auto',
    marginBottom: 'auto',
  },
});

export default Timer;
