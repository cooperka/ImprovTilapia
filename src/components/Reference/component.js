import Immutable from 'immutable';
import React, { Component } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { inject, observer } from 'mobx-react/native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { List, TouchableRipple } from 'react-native-paper';

import { references } from './utils';
import FloatingNav from '../FloatingNav/component';
import { color } from '../../constants';

/** Adapted from Paper's `List.Icon`. */
const getSectionIcon = (name) => (props) => {
  const { color: iconColor } = props;

  return (
    <View style={styles.drawerIcon} pointerEvents="box-none">
      <MaterialCommunityIcons name={name} size={24} color={iconColor} />
    </View>
  );
};

function ListItem({ title, description, expanded, onPress }) {
  return (
    <TouchableRipple style={styles.itemContainer} onPress={onPress}>
      <View style={styles.item}>
        {title ? <Text style={styles.itemTitle}>{title}</Text> : null}
        {description ? (
          <Text
            numberOfLines={expanded ? undefined : 1}
            style={styles.itemDescription}
          >
            {description}
          </Text>
        ) : null}
      </View>
    </TouchableRipple>
  );
}

@inject('referenceSettings')
@observer
class Reference extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: 'Reference',
    headerLeft: () => <FloatingNav navigation={navigation} />,
  });

  static tabNavigationOptions = () => ({
    tabBarIcon: ({ tintColor }) => (
      <MaterialCommunityIcons
        name="book-open-page-variant"
        color={tintColor}
        size={20}
      />
    ),
  });

  state = {
    expandedItems: Immutable.Set(),
  };

  handleToggleItem = (itemName) => {
    const { expandedItems } = this.state;
    const exists = expandedItems.contains(itemName);
    const newItems = exists
      ? expandedItems.delete(itemName)
      : expandedItems.add(itemName);
    this.setState({ expandedItems: newItems });
  };

  render() {
    const { expandedItems } = this.state;

    return (
      <View style={styles.container}>
        {references.map(({ name: sectionName, iconName, items }) => (
          <List.Accordion
            key={sectionName}
            title={sectionName}
            left={getSectionIcon(iconName)}
            expanded={true}
          >
            {items.map(({ name: itemName, description }) => (
              <ListItem
                key={itemName}
                title={itemName}
                expanded={expandedItems.contains(itemName)}
                description={description}
                onPress={() => this.handleToggleItem(itemName)}
              />
            ))}
          </List.Accordion>
        ))}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  /** Adapted from Paper's `List.Icon`. */
  drawerIcon: {
    margin: 8,
    height: 40,
    width: 40,
    alignItems: 'center',
    justifyContent: 'center',
  },
  itemContainer: {
    marginTop: 8,
    marginBottom: 8,
  },
  /** Adapted from Paper's `List.ListAccordion`. */
  item: {
    marginLeft: 64 + 8,
    marginRight: 16,
  },
  /** Adapted from Paper's `List.ListItem`. */
  itemTitle: {
    color: color.MATERIAL_PRIMARY,
    fontSize: 16,
  },
  itemDescription: {
    color: color.MATERIAL_SECONDARY,
    fontSize: 14,
    marginBottom: 8,
  },
});

export default Reference;
