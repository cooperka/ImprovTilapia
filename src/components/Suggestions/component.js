import React, { Component } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { inject, observer } from 'mobx-react/native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { Col, Row, Grid } from 'react-native-easy-grid';

import FloatingNav from '../FloatingNav/component';
import { Button } from 'react-native-paper';
import { suggestionTypes, suggestions } from './utils';
import { Constants } from 'expo';

// Organized in pairs (left and right column).
const suggestionOrder = [
  suggestionTypes.RELATIONSHIP,
  suggestionTypes.EMOTION_SHORT,

  suggestionTypes.OCCUPATION,
  suggestionTypes.EMOTION_FULL,

  suggestionTypes.LOCATION,
  suggestionTypes.EMOTIONAL_DIAD,
];

@inject('suggestionsSettings')
@observer
class Suggestions extends Component {
  static navigationOptions = ({ navigation }) => ({
    tabBarIcon: ({ tintColor }) => (
      <MaterialCommunityIcons
        name="message-outline"
        color={tintColor}
        size={20}
      />
    ),
    tabBarVisible: !navigation.getParam('isFullscreen'),
  });

  state = {
    currCategory: null,
    currSuggestion: 'Press a button below to get a suggestion',
    width: null,
  };

  handleLayoutChange = ({
    nativeEvent: {
      layout: { width, height },
    },
  }) => {
    const { navigation } = this.props;
    navigation.setParams({ isFullscreen: width > height });

    this.setState({ width });
  };

  handleNewSuggestion = (category, suggestionFn) => () => {
    this.setState({ currCategory: category, currSuggestion: suggestionFn() });
  };

  isActive = (name) => {
    const { currCategory } = this.state;
    return name === currCategory;
  };

  render() {
    const { navigation } = this.props;
    let { currSuggestion, width } = this.state;
    const textLengthFactor =
      currSuggestion.length > 20
        ? // 5% smaller for each character beyond 20, but at most half the normal size.
          Math.min(((currSuggestion.length - 20) * 5) / 100 + 1, 2)
        : 1;

    return (
      <View style={styles.container} onLayout={this.handleLayoutChange}>
        <FloatingNav navigation={navigation} />

        <View style={styles.suggestionContainer}>
          <Text
            style={[
              styles.suggestion,
              { fontSize: width / 8.0 / textLengthFactor },
            ]}
          >
            {currSuggestion}
          </Text>
        </View>

        <Grid style={styles.buttonsContainer}>
          {suggestionOrder
            // Get data.
            .map((key) => suggestions[key])
            // Pair into columns.
            .reduce((reduction, value, index) => {
              if (index % 2 === 0) {
                reduction.push([value]);
              } else {
                reduction[reduction.length - 1].push(value);
              }
              return reduction;
            }, [])
            // Render each pair.
            .map((pair, index) => (
              <Row key={index}>
                {pair.map(({ name, getRandomThing }) => (
                  <Col key={name}>
                    <Button
                      mode={this.isActive(name) ? 'contained' : 'text'}
                      style={styles.button}
                      onPress={this.handleNewSuggestion(name, getRandomThing)}
                    >
                      {name}
                    </Button>
                  </Col>
                ))}
              </Row>
            ))}
        </Grid>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
  suggestionContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  suggestion: {
    marginTop: Constants.statusBarHeight,
    marginBottom: 8,
    textAlign: 'center',
    color: '#0f0',
    marginHorizontal: 16,
  },
  buttonsContainer: {
    flex: 1,
    alignItems: 'center',
    maxWidth: 500,
    maxHeight: 140,
    marginBottom: 16,
  },
  button: {
    marginHorizontal: 16,
  },
});

export default Suggestions;
