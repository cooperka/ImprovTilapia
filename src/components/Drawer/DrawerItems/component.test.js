import React from 'react';
import { shallow } from 'enzyme';

import Component, {
  RouteItems,
  TimerItems,
  ReferenceItems,
  SuggestionsItems,
} from './component';
import { TimerSettingsModel } from '../../Timer/model';
import { SuggestionsSettingsModel } from '../../Suggestions/model';
import { ReferenceSettingsModel } from '../../Reference/model';

const mockStores = {
  timerSettings: new TimerSettingsModel(),
  suggestionsSettings: new SuggestionsSettingsModel(),
  referenceSettings: new ReferenceSettingsModel(),
};

jest.mock('../utils', () => ({
  getCurrRouteName: () => 'mockRoute',
}));

it('renders as expected (root)', async () => {
  const wrapper = shallow(<Component navigation={{}} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders as expected (route items)', async () => {
  const wrapper = shallow(<RouteItems />);
  expect(wrapper).toMatchSnapshot();
});

it('renders as expected (timer items)', async () => {
  const wrapper = shallow(<TimerItems {...mockStores} currRouteName={''} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders as expected (suggestions items)', async () => {
  const wrapper = shallow(
    <SuggestionsItems {...mockStores} currRouteName={''} />,
  );
  expect(wrapper).toMatchSnapshot();
});

it('renders as expected (reference items)', async () => {
  const wrapper = shallow(
    <ReferenceItems {...mockStores} currRouteName={''} />,
  );
  expect(wrapper).toMatchSnapshot();
});
