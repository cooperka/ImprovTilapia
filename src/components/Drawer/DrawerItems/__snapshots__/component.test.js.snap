// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<Accordion
  left={[Function]}
  title="Timer settings"
>
  <Item
    title="First item"
  />
  <Item
    title="Second item"
  />
</Accordion>
`;
